### PR TITLE
Fix/982 file system exception no free space

### PR DESCRIPTION
**Pull request by YassinNouh21**
📝 **Summary**
---
**This PR addresses in device has no  file system exception no free space**

**Description**
---
This PR introduces several key enhancements to the app:

#### 1. Add new class of `DeviceInfo` and `DeviceManagerProvider`:
- Implemented the `DeviceInfo` class to encapsulate essential device-specific information such as device ID, brand, model, OS version, app version, and storage space.
- Developed `DeviceManagerProvider`, extending `AsyncNotifier`, to manage and asynchronously provide device-related information throughout the app.
- Integrated methods in `DeviceManagerProvider` to fetch various pieces of device information.

#### 2. Integration of Dio for HTTP Requests:
- Added `DioModule` class as a configurable wrapper for the Dio HTTP client, which includes custom settings such as base URL, headers, timeouts, and interceptors.
- Introduced `dioProvider` as a Riverpod provider for creating DioModule instances with custom configurations, enhancing the flexibility of our HTTP client setup.

#### 3. Cache Interceptor Support:
- Implemented `CacheInterceptorHelper` to manage caching policies for Dio requests, which will help in efficient network management and reduced data usage.
- The caching policy is adjusted based on the device's storage state, ensuring optimal performance even in low-storage scenarios.

#### 4. Enhancements in the Splash Screen:
- Refactored the Splash screen to extend `ConsumerStatefulWidget`, allowing better state management with Riverpod.
- Added a new `ErrorState` for storage errors and integrated a mechanism in the Splash screen to handle and display storage-related errors.
- Add handling scenarios where the device has insufficient storage.

**Note**
- The features "Cache Interceptor" and "Integration of Dio for HTTP Requests" have been conceptualized and partially developed, but they are not yet integrated into the current codebase. These components are for future updates. 

**Tests**
---
🧪 **Use case 1:  Show exception when the storage is full**
---

https://github.com/mawaqit/android-tv-app/assets/70436855/24d090ba-6ce1-4041-add8-3f6c1abd1386

---
🧪 **Use case 2: When press to recheck the storage size**
---

https://github.com/mawaqit/android-tv-app/assets/70436855/4daf86ff-4722-4418-92a3-b6c6807bc896

🧪 **Use case 3: When press to recheck the storage size and it is still full**
---

https://github.com/mawaqit/android-tv-app/assets/70436855/90a75335-40ee-4b9b-a90c-14cd14139320

**Checklist:**
---
- [x] **Coding Standards:** I have reviewed my code to ensure it follows the project's coding standards.
- [x] **Testing:** I have tested the changes and they work as expected.
- [x] **Merge Conflicts:** I have resolved any merge conflicts with the latest main/development branch.
- [x] **Branch Status:** The branch is up-to-date with the target branch (main/development).

### CODE REVIEW

File: lib/src/helpers/device_manager_provider.dart [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+      totalSpace: totalSpace,
+      freeSpace: freeSpace,
+    );
+  }
+
+  /// [getFreeSpace] Asynchronously gets the free disk space and updates the state.
+  Future<void> getFreeSpace() async {
+    state = const AsyncLoading();
+    state = await AsyncValue.guard(() async {
+      final freeSpace = await DiskSpace.getFreeDiskSpace;
```
Comment (ghassenbenzahra123) [inline lib/src/helpers/device_manager_provider.dart:59]: Please handle null case of freespace

File: lib/main.dart [hunk 2]
```diff
@@ -3,7 +3,7 @@ import 'package:flutter/material.dart';
 import 'package:flutter/services.dart';
 import 'package:flutter_localizations/flutter_localizations.dart';
 import 'package:flutter_riverpod/flutter_riverpod.dart'
-    show ProviderBase, ProviderContainer, ProviderObserver, ProviderScope;
+    show AsyncValueX, ConsumerWidget, FutureProvider, ProviderBase, ProviderContainer, ProviderObserver, ProviderScope, UncontrolledProviderScope;
```
Comment (ghassenbenzahra123) [inline lib/main.dart:6]: remove unused imports

File: lib/main.dart [hunk 3]
```diff
@@ -14,27 +14,44 @@ import 'package:mawaqit/src/helpers/Api.dart';
 import 'package:mawaqit/src/helpers/AppRouter.dart';
 import 'package:mawaqit/src/helpers/ConnectivityService.dart';
 import 'package:mawaqit/src/helpers/CrashlyticsWrapper.dart';
+import 'package:mawaqit/src/helpers/device_manager_provider.dart';
+import 'package:mawaqit/src/pages/ErrorScreen.dart';
```
Comment (ghassenbenzahra123) [inline lib/main.dart:18]: this one also its unused

**Review by ghassenbenzahra123 — COMMENTED** [review]
Manually tested all the cases works good missing minor modifications. 
Great work Yassin

**Comment by ibrahim-zehhaf-mawaqit** [comment]
In 🧪 Use case 1: Show exception when the storage is full
I see that the screen is online, as we disccused, we show the error screen only if there no space and the app is offline

File: pubspec.yaml [hunk 4]
```diff
@@ -110,6 +110,9 @@ dependencies:
   collection: any
   disk_space: ^0.2.1
 
+  # alerting library
+  flash: ^3.0.5+2
```
Comment (ibrahim-zehhaf-mawaqit) [inline pubspec.yaml:114]: you need this library ? now you don't show any alert right ?

File: lib/src/helpers/cache_interceptor.dart [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+  DioCacheInterceptor _createCacheInterceptor(CachePolicy policy, String path) {
+    return DioCacheInterceptor(
+      options: CacheOptions(
+        store: HiveCacheStore(
+          path == '' ? null : path,
+          hiveBoxName: 'mawaqit',
+        ),
+        policy: policy,
+        keyBuilder: CacheOptions.defaultCacheKeyBuilder,
+        hitCacheOnErrorExcept: [401, 403],
```
Comment (ibrahim-zehhaf-mawaqit) [inline lib/src/helpers/cache_interceptor.dart:32]: what about 5XX statuses

**Comment by ibrahim-zehhaf-mawaqit** [comment]
> @ibrahim-zehhaf-mawaqit if the package for some reason can't access the disk to read the storage. Should I return error or not ?

No just handle the exception and catch it to avoid flooding in sentry

File: lib/src/helpers/cache_interceptor.dart (hunk 5 above)
Comment (ibrahim-zehhaf-mawaqit) [reply lib/src/helpers/cache_interceptor.dart:38]: up

**Comment by ibrahim-zehhaf-mawaqit** [comment]
@ghassenbenzahra123 Need approve

**Comment by ibrahim-zehhaf-mawaqit** [comment]
Any update about this ?

**Comment by ibrahim-zehhaf-mawaqit** [comment]
@YassinNouh21 please fix the conflicts

**Review by ghassenbenzahra123 — COMMENTED** [review]
After Manually testing the branch:  

- the exception FileSystemException no longer appears. But a new, non-blocking of the fix, exception raises here you can find the full stack trace it happens when the free space is less then 200MB and at each launch of the app. You can check the [issue ](https://mawaqit-01.sentry.io/issues/4970236586/?environment=debug&project=4504947427901440&query=is%3Aunresolved&referrer=issue-stream&statsPeriod=1h&stream_index=1&utc=true) on sentry for the stack trace.
- Changing the mosque from the settings does not apply directly you need to restart the app, to see the changes.
-  The flash warning message appears multiple times you can see in the recording.

[flashMessage.webm](https://github.com/mawaqit/android-tv-app/assets/35707318/403a4c1b-cbd2-42e3-8580-8878d3d94d55)

**Comment by ibrahim-zehhaf-mawaqit** [comment]
@YassinNouh21 please make this your priority

**Pull request closed without merging**